src/type.js

```diff
@@ -736,3 +701,49 @@ Sk.builtin.type.prototype["__format__"] = function(self, format_spec) {
 };
 
 Sk.builtin.type.pythonFunctions = ["__format__"];
+
+Sk.builtin.type.$shortcutDunder = function (klass, skulpt_name, magic_name, magic_func, canSuspendIdx) {
+    klass.prototype[skulpt_name] = function () {
+        var canSuspend = false;
+        var len = arguments.length;
+        var args, i, j;
+        if ((canSuspendIdx !== null) && (canSuspendIdx <= len)) {
+            args = new Array(len);
+        } else {
+            args = new Array(len+1);
+        }
+
+        args[0] = this;
+        j = 1;
+        for (i = 0; i < len; i++) {
+            if (i === (canSuspendIdx-1)) {
+                canSuspend = arguments[i];
+            } else {
+                args[j] = arguments[i];
+                j += 1;
+            }
+        }
+
+        if (canSuspend) {
+            return Sk.misceval.callsimOrSuspendArray(magic_func, args);
+        } else {
+            return Sk.misceval.callsimArray(magic_func, args);
+        }
```

Comment:
This is a lot of code on a hot path (it runs every time the slot function is called). Most (in fact, nearly all) of these decisions can be hoisted up in to `$shortcutDunder`, so the branches only need to be taken once.

Reply:
I'm not particularly confident on suspensions yet. But I separated them into two functions. 
one for magic methods that can suspend and one for magic methods that can't...